File: apps/api/plane/app/serializers/state.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Module imports
from .base import BaseSerializer
from rest_framework import serializers

from plane.db.models import State, StateGroup, SubState


class SubStateLiteSerializer(BaseSerializer):
    class Meta:
        model = SubState
        fields = ["id", "state_id", "project_id", "workspace_id", "name", "color", "icon", "sequence"]
        read_only_fields = fields


class SubStateSerializer(BaseSerializer):
    class Meta:
        model = SubState
        fields = ["id", "project_id", "workspace_id", "state_id", "name", "color", "icon", "sequence"]
        read_only_fields = ["project_id", "workspace_id", "state_id"]


class StateSerializer(BaseSerializer):
    order = serializers.FloatField(required=False)
    sub_states = SubStateLiteSerializer(source="state_sub_states", many=True, read_only=True)

    class Meta:
        model = State
        fields = [
            "id",
            "project_id",
            "workspace_id",
            "name",
            "color",
            "group",
            "default",
            "description",
            "sequence",
            "order",
            "sub_states",
        ]
        read_only_fields = ["workspace", "project"]

    def validate(self, attrs):
        if attrs.get("group") == StateGroup.TRIAGE.value:
            raise serializers.ValidationError("Cannot create triage state")
        return attrs


class StateLiteSerializer(BaseSerializer):
    class Meta:
        model = State
        fields = ["id", "name", "color", "group"]
        read_only_fields = fields

File: packages/types/src/state.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TStateGroups = "backlog" | "unstarted" | "started" | "completed" | "cancelled";

export interface ISubState {
  readonly id: string;
  state_id: string;
  project_id: string;
  workspace_id: string;
  name: string;
  color: string;
  icon: string | null;
  sequence: number;
}

export interface IState {
  readonly id: string;
  color: string;
  default: boolean;
  description: string;
  group: TStateGroups;
  name: string;
  project_id: string;
  sequence: number;
  workspace_id: string;
  order: number;
  sub_states?: ISubState[];
}

export interface IStateLite {
  color: string;
  group: TStateGroups;
  id: string;
  name: string;
}

export interface IStateResponse {
  [key: string]: IState[];
}

export type TStateOperationsCallbacks = {
  createState: (data: Partial<IState>) => Promise<IState>;
  updateState: (stateId: string, data: Partial<IState>) => Promise<IState | undefined>;
  deleteState: (stateId: string) => Promise<void>;
  createSubState: (stateId: string, data: Partial<ISubState>) => Promise<ISubState>;
  updateSubState: (stateId: string, subStateId: string, data: Partial<ISubState>) => Promise<ISubState | undefined>;
  deleteSubState: (stateId: string, subStateId: string) => Promise<void>;
  moveStatePosition: (stateId: string, data: Partial<IState>) => Promise<void>;
  markStateAsDefault: (stateId: string) => Promise<void>;
};
